R/graded.R

```diff
@@ -90,3 +220,29 @@ legacy_graded <- function(...) {
     graded(...)
   )
 }
+
+get_from_env <- function(x, env, .caller) {
+  tryCatch(
+    get(x, envir = env),
+    error = function(e) {
+      label <- get0(".label", env, ifnotfound = NULL)
+      label <- if (!is.null(label)) paste0("In exercise `", label, "`: ")
+      not_found <- sprintf(
+        gettext("object %s not found", domain = "R"), 
+        paste0("'", x, "'")
+      )
+      if (grepl(not_found, e$message, fixed = TRUE)) {
```

Comment:
I think the approach above breaks down if the value `x` actually equals `NULL`.

Maybe we could approach this with a mix of `get0()` and `rlang`. 

```r
val <- get0(x, envir = env, ifnotfound = rlang::missing_arg())
if (!rlang::is_missing(val)) return(val)
```

Reply:
If `x` is `NULL`, won't `get(x, envir = env)` return cleanly?

```
env <- new.env()
get("val", envir = env)
#> Error in get("val", envir = env): object 'val' not found
env$val <- NULL
get("val", envir = env)
#> NULL
```

Reply:
I guess I was trying to get away from a tryCatch and trying to deconstruct the message.

Do you know if there are other types of errors that can be thrown besides a "object X not found"?

Reply:
I suppose we could instead use `exists(x, envir = env)` and if it doesn't exist return a `missing_arg()` to have `pass_if_equal()` or `fail_if_equal()` handle it.